[skip-CI][NFC][ntuple] describe code architecture (#14505)
root-project/root: tree/ntuple/v7/doc/architecture.md

General Principles
------------------

The RNTuple classes provide the functionality to read, write, and describe RNTuple datasets.
The core classes, such as `RNTupleReader` and `RNTupleWriter`, are part of the RNTuple library.
Additional tools, such as the `RNTupleImporter` and the `RNTupleInspector`, are part of the RNTupleUtils library,
which depends on the RNTuple library.

The RNTuple classes are organized in layers:
the storage layer, the primitives layer, the logical layer and the event iteration layer.
Most classes in the storage layer and the primitives layer are in the `ROOT::Internal` namespace,
with the notable exception of the descriptor classes (`RNTupleDescriptor`, `RFieldDescriptor`, etc.).
Most classes in the upper layers provide public interfaces.

| Layer      | Description                                                         | Example of classes                                          |
|------------|---------------------------------------------------------------------|-------------------------------------------------------------|

> | Storage    | Read and write pages (physical: file, DAOS; virtual: e.g. buffered) | RPage{Source,Sink}, RNTupleDescriptor, RClusterPool         |

Reviewer: ```suggestion
| Storage    | Read and write pages (physical: file, object store; virtual: e.g. buffered) | RPage{Source,Sink}, RNTupleDescriptor, RClusterPool         |
```

| Primitives | Storage-backed columns of simple types                              | RColumn, RColumnElement, RPage                              |
| Logical    | Mapping of C++ types onto columns                                   | RField, RNTupleModel, REntry                                |
| Iteration  | Reading and writing events / properties                             | RNTuple{Reader,Writer}, RNTupleView, RNTupleDS (RDataFrame) |
| Tooling    | Higher-level, RNTuple related utility classes                       | RNTupleMerger, RNTupleImporter, RNTupleInspector            |

The RNTuple classes are, unless explicitly stated otherwise, conditionally thread safe.

The read and write APIs provide templated, compile-time type-safe APIs,
APIs where the type at hand is passed as string and which are runtime type-safe,
and type-unsafe APIs using void pointers.

On I/O errors and invalid input, RNTuple classes throw an `RException`.

Walkthrough: Reading Data
-------------------------

```c++
auto file = TFile::Open("data.root");

> auto ntuple = file->Get<RNTuple>("ntpl");

Reviewer: Do we want to actually show unsafe APIs in this document? e.g. in this very same snippet the `TFile *` is not deleted explicitly and only works because of the global list of cleanups

// Option 1: entire row
// The reader creates a page source; the page source creates a model from the on-disk information
auto reader = RNTupleReader::Open(ntuple);
// Populate the objects that are used in the model's default entry
reader->LoadEntry(0);
std::shared_ptr<float> pt = reader->GetDefaultEntry().GetPtr<float>("pt");

// Option 2: imposed model
auto model = RNTupleModel::Create();
auto pt = model->MakeField<float>("pt");
// The reader checks the passed model for compatibility; only the subset of fields defined in the model is read
auto reader = RNTupleReader::Open(std::move(model), ntuple);
reader->LoadEntry(0);

> // Option 3: through views

Reviewer: ```suggestion
// Option 3: through views
// Each view will only trigger reading of the related field, without reading other fields at the same entry number.
```

### REntry
The REntry represents a row/entry in an RNTuple.
It contains a list of `RValue` objects that correspond to the top-level fields of the originating model.
The entry gives access to the shared pointers corresponding to the top-level fields.
It also provides functionality to bind application-provided pointers.

An REntry can be passed to `RNTupleWriter::Fill()` and `RNTupleReader::LoadEntry()`.
Otherwise, the reader/writer uses the default entry of its model.

An entry can safely outlive its originating model.
New objects cannot anymore be created (`EmplaceNewValue` will throw an exception), but the entry is still properly destructed.

### RNTupleWriter, RNTupleParallelWriter
The RNTupleWriter is the primary interface to create an RNTuple.
The writer takes ownership of a given model.
The writer can either add an RNTuple to an existing ROOT file (`RNTupleWriter::Append()`) or create a new ROOT file with an RNTuple (`RNTupleWriter::Recreate()`).
Once created, entries are added to an RNTuple either serially (`RNTupleWriter::Fill()`) or in concurrently in multiple threads with the `RNTupleParallelWriter`.

> Once committed (e.g. by releasing the RNTupleWriter), the RNTuple is immutable and cannot be ammended.

Reviewer: ```suggestion
Once committed (e.g. by releasing the RNTupleWriter), the RNTuple is immutable and cannot be amended.
```

An RNTuple that is currently being written cannot be read.

### RNTupleReader
The RNTupleReader is the primary interface to read and inspect an RNTuple.
An RNTupleReader owns a model: either a model created from the on-disk information or an imposed, user-provided model.
The user-provided model can be limited to a subset of fields.
Data is populated to an explicit `REntry` or the model's default entry through `RNTupleReader::LoadEntry()`.

The reader can create `RNTupleView` objects for the independent reading of individual fields.
The reader can create `RBulk` objects for bulk reading of individual fields.

Additionally, the reader provides access to a cached copy of the descriptor.
It can display individual entries (`RNTupleReader::Show()`) and summary information (`RNTupleReader::PrintInfo()`).

### RNTupleView<T>
RNTuple views provide read access to individual fields.
Views are created from an RNTupleReader.

> Views are templated; for simple types (e.g., `float`, `int`), views provide read-only access directly to an RNTuple page in memory.

Reviewer: Unless the user provides a memory location for it after https://github.com/root-project/root/pull/14849 . But we can amend the document if/when that PR is merged

The local file implementation on Linux uses uring for vector reads, if available.
A raw file that uses a TFile (to be implemented) provides access to the full set of TFile implementations, e.g. TMemFile.

Tooling
-------

### RNTupleMerger
The `RNTupleMerger` is an internal class and part of the core RNTuple library.
It concatenates RNTuple data from several sources into a combined sink.
It implements "fast merging", i.e. copy-based merging that does not decompress and recompress pages.
The RNTupler merger is used by the `TFileMerger` and thus provides RNTuple merge support in `hadd` and `TBufferMerger`.

### RNTupleImporter
The RNTupleImporter creates RNTuple data sets from ROOT trees.
It is part of the `ROOTNTupleUtil` library.

### RNTupleInspector
The RNTupleInspector provides insights of an RNTuple, e.g. the distribution of data volume wrt. column types.
It is part of the `ROOTNTupleUtil` library.

Ownership Model
---------------

By default, objects involved in RNTuple I/O (objects read from disk or written to disk) are passed to RNTuple as shared pointers.
Both RNTuple or the application may create the object.

> Raw pointers to objects can be passed to RNTuple -- such objects are considered as owned by the application with a sufficiently long lifetime.

Reviewer: Maybe be more explicit and add that is the caller's responsibility to guarantee the lifetime of the objects

> This document is meant for ROOT developers. It provides background information on the RNTuple code design and behavior.

> > The RNTuple code uses the nomenclature from the RNTuple format specification (e.g. "field", "column", "anchor", etc.).

Reviewer 2: ```suggestion
> The RNTuple code uses the nomenclature from the [RNTuple format specification](https://github.com/root-project/root/blob/master/tree/ntuple/v7/doc/specifications.md) (e.g. "field", "column", "anchor", etc.).
```

General Principles
------------------

The RNTuple classes provide the functionality to read, write, and describe RNTuple datasets.
The core classes, such as `RNTupleReader` and `RNTupleWriter`, are part of the RNTuple library.
Additional tools, such as the `RNTupleImporter` and the `RNTupleInspector`, are part of the RNTupleUtils library,

> that depends on the RNTuple library.

Reviewer 2: ```suggestion
which depends on the RNTuple library.
```

The RNTuple classes are organized in layers:

> the storage layer, the primitives layer, the logical layer and the event iteration.

Reviewer 2: ```suggestion
the storage layer, the primitives layer, the logical layer and the event iteration layer.
```

Most classes in the storage layer and the primitives layer are in the `ROOT::Internal` namespace,
with the notable exception of the descriptor classes (`RNTupleDescriptor`, `RFieldDescriptor`, etc.).
Most classes in the upper layers provide public interfaces.

| Layer      | Description                                                         | Example of classes                                   |
|------------|---------------------------------------------------------------------|------------------------------------------------------|
| Storage    | Read and write pages (physical: file, DAOS; virtual: e.g. buffered) | RPage{Source,Sink}, RNTupleDescriptor, RClusterPool  |
| Primitives | Storage-backed columns of simple types                              | RColumn, RColumnElement, RPage                       |
| Logical    | Mapping of C++ types onto columns                                   | RField, RNTupleModel, REntry                         |

> | Iteration  | Reading and writing events / properties                             | RNTuple{Reader,Writer}, RNTupleView, RNTupleDS (RDF) |

Reviewer 2: ```suggestion
| Iteration  | Reading and writing events / properties                             | RNTuple{Reader,Writer}, RNTupleView, RNTupleDS (RDataFrame) |
```

| Tooling    | Higher-level, RNTuple related utility classes                       | RNTupleMerger, RNTupleImporter, RNTupleInspector     |

The RNTuple classes are, unless explicitly stated otherwise, conditionally thread safe.

The read and write APIs provide templated, compile-time type-safe APIs,
APIs where the type at hand is passed as string and which are runtime type-safe,
and type-unsafe APIs using void pointers.

> On I/O errors and invalid input, RNTuple classes throw an `RException`.

Reviewer 2: This is more for future reference/note to self, but we should check if this is the case everywhere (IIRC there still are a few places, accessible through public APIs, with `asserts`).

Walkthrough: Reading Data
-------------------------

```c++
auto file = TFile::Open("data.root");
auto ntuple = file->Get<RNTuple>("ntpl");

// Option 1: entire row
// The reader creates a page source; the page source creates a model from the on-disk information
auto reader = RNTupleReader::Open(ntuple);
// Populate the objects that are used in the model's default entry
reader->LoadEntry(0);
std::shared_ptr<float> pt = reader->GetDefaultEntry().GetPtr<float>("pt");

// Option 2: imposed model
auto model = RNTupleModel::Create();

> model->MakeField<float>("pt");

Reviewer 3: Maybe show that this can be used to access the value:
```suggestion
auto pt = model->MakeField<float>("pt");
```
(IIRC)

// The reader checks the passed model for compatibility; only the subset of fields defined in the model is read
auto reader = RNTupleReader::Open(std::move(model), ntuple);
reader->LoadEntry(0);

// Option 3: through views
auto reader = RNTupleReader::Open(ntuple);

> auto viewPt = reader->GetView("pt");

Reviewer 3: should be
```suggestion
auto viewPt = reader->GetView<float>("pt");
```

// Load the pt from the first entry
auto pt = viewPt(0);
```

In the above cases, RNTuple creates the objects being read into.
It is also possible to bind already existing objects.

> This is shown for entries and works similarly for views.

Reviewer 2: ```suggestion
This is shown below for entries and works similarly for views.
```

```c++
// A bare entry is an entry that has initially no bindings (all top-level fields need to be bound by the caller)
auto entry = reader->GetModel().CreateBareEntry();
auto ptToken = entry->GetToken("pt");

// Option 1: type safe, shared ownership
std::shared_ptr<float> ptTypedSharedPtr;
entry->BindValue(ptToken, ptTypedSharedPtr);

// Option 2: type unsafe, shared ownership
std::shared_ptr<void> ptVoidSharedPtr;
entry->BindValue(ptToken, ptVoidSharedPtr);

// Option 3: type unsafe, application owns the object
void *ptVoidPtr;
entry->BindRawPtr(ptToken, ptVoidPtr);

// Option 4: switch back from application-provided object to RNTuple-created object
entry->EmplaceNewValue(ptToken);

// For all options: use an explicit entry
reader->LoadEntry(0, *entry);
```

Walkthrough: Writing Data
-------------------------

```c++
auto model = RNTupleModel::Create();
// Add a field to the model and return the shared pointer for that field in the model's default entry.

> auto ptrPt = model->MakeField<float>(pt);

Reviewer 2: ```suggestion
auto ptrPt = model->MakeField<float>("pt");
```

auto file = TFile::Open("data.root", "APPEND");
// The writer creates a page sink and connects the model's fields to it
auto writer = RNTupleWriter::Append(std::move(model), "ntpl", *file);

> ptrPt = 1.0;

Reviewer 3: ```suggestion
*ptrPt = 1.0;
```

The meta-data include the schema (fields and columns), information about clusters and the page locations.
The descriptor classes are closely related to the format specification.

For normal read and write tasks, access to the descriptor is not necessary.
One notable exception is bulk reading, where the descriptor can be used to determine entry boundaries of clusters.
The descriptors are used internally, e.g. to build an RNTupleModel from the on-disk information.
The descriptors are also useful for inspection purposes.

The descriptor classes contain a copy of the meta-data; they are not linked to an open page source.
A descriptor can be used after its originating page source has been deleted.

### RField<T>
The RField<T> classes are central in RNTuple:
they link the in-memory representation of data types to their on-disk representation.
All field classes inherit from `RFieldBase`.

Every type with RNTuple I/O supported has a corresponding RField<T> template specialization.
Complex types are composed of fields and sub fields.
E.g., a `struct` is represented by a parent field for the `struct` itself and a subfield for every member of the `struct`.

> Fields of complex types have type-erased versions in addition to the templated ones (e.g., `RVectorField`, `RClassField`).

Reviewer 2: ```suggestion
Fields of complex types have type-erased versions in addition to the templated ones (e.g., `RCollectionField`, `RClassField`).
```

Author: I think here I would like to stick to the `RVectorField`. The `RCollectionField` is special in that it makes the connection between an RNTupleCollectionWriter and the RNTupleWriter.

In this way, fields can be constructed even if the type information is only available at runtime.
To this end, `RFieldBase::Create()` creates an `RField` object from a type string.

On the "in-memory" side, fields can construct and destroy objects of their wrapped type
(cf. `CreateValue()`, `CreateObject()`, `GetDeleter()` methods).
Existing objects in memory can be bound to fields (cf. `BindValue()` method).

On the "on-disk" side, fields know about the possible column representations of their wrapped type.
Upon connecting a field to a page source or page sink,
fields create `RColumn` objects and register them with the page source/sink.
When reading and writing data, the field maps the in-memory information of an object to read/write calls on its columns.
For instance, when writing a `std::vector<float>`,

> the field writes to an index column (storing information about the size of the vector), and to a float column (the actual values).

Reviewer 3: not sure if important here: the `vector` field has a subfield for the `float`s which in turn handles the column.

During its lifetime, a field undergoes the following possible state transitions:
```
[*] --> Unconnected --> ConnectedToSink ----
             |      |                      |
             |      --> ConnectedToSource ---> [*]
             |                             |
             -------------------------------
```

The RField class hierarchy is fixed and not meant to be extended by user classes.

### RField::RValue
The `RValue` class makes the connection between an object in memory and the corresponding field used for I/O.
It contains a shared pointer of the object, i.e. RNTuple and the application share ownership of objects.
The object in an RValue can either be created by an RNTuple field (cf. `RField<T>::CreateValue()` method)

> or provided by the application (cf. `RField<T>::BindValue` method).

Reviewer 2: ```suggestion
or provided by the application (cf. `RField<T>::BindValue()` method).
```

Once created, entries are added to an RNTuple either serially (`RNTupleWriter::Fill()`) or in concurrently in multiple threads with the `RNTupleParallelWriter`.
Once committed (e.g. by releasing the RNTupleWriter), the RNTuple is immutable and cannot be appended.
An RNTuple that is currently being written cannot be read.

### RNTupleReader
The RNTupleReader is the primary interface to read and inspect an RNTuple.
An RNTupleReader owns a model: either a model created from the on-disk information or an imposed, user-provided model.
The user-provided model can be limited to a subset of fields.
Data is populated to an explicit `REntry` or the model's default entry through `RNTupleReader::LoadEntry()`.

The reader can create `RNTupleView` objects for the independent reading of individual fields.
The reader can create `RBulk` objects for bulk reading of individual fields.

Additionally, the reader provides access to a cached copy of the descriptor.
It can display individual entries (`RNTupleReader::Show()`) and summary information (`RNTupleReader::PrintInfo()`).

> ### RNTupleView

Reviewer 3: ```suggestion
### RNTupleView<T>
```

> ### RNTupleView

Reviewer 2: Perhaps make a mention of their lifetime?

RNTuple views provide read access to individual fields.
Views are created from an RNTupleReader.
Views are templated; for simple types (e.g., `float`, `int`), views provide read-only access directly to an RNTuple page in memory.
Complex types and void views require additional memory copies to populate an object in memory from the column data.

A view can iterate over the entry range, over the field range, and over the range of a collection within an entry.
For instance, for a field `std::vector<float> pt`, a view can iterate over all `pt` values of all entries, or over the `pt` values of a particular entry.

Internal Classes
----------------

### RNTupleDS
The `RNTupleDS` class is an internal class that provides an RNTuple data source for RDataFrame.

> It is part of the dataframe library.

Reviewer 2: ```suggestion
It is part of the `ROOTDataFrame` library.
```

The RNTuple data source supports chains with a constructor that takes a list of input files.
The RNTuple data source also supports multi-threaded dataframes, parallelized on the file and cluster level.

The data source exposes inner fields of complex collections.
For instance, if the data model contains a vector of `Event` classes, where each `Event` has `pt` and `eta` floats,

> the data frame can use the event vector itself (`Event` column) as well as the `float` columns `Event.pt` and `Event.eta`.

Reviewer 2: ```suggestion
the dataframe can use the event vector itself (`Event` column) as well as the `float` columns `Event.pt` and `Event.eta`.
```

For writing data, the RMiniFile can either use a proper `TFile` (descendant) or a C file stream (only for new ROOT files with a single RNTuple).
For reading, the `RMiniFile` always uses an `RRawFile`.

### RRawFile
The RRawFile internal abstract class provides an interface to read byte ranges from a file, including vector reads.
Concrete implementations exist for local files, XRootD and HTTP (the latter two through the ROOT plugin mechanism).
The local file implementation on Linux uses uring for vector reads, if available.
A raw file that uses a TFile (to be implemented) provides access to the full set of TFile implementations, e.g. TMemFile.

Tooling
-------

### RNTupleMerger
The `RNTupleMerger` is an internal class and part of the core RNTuple library.
It concatenates RNTuple data from several sources into a combined sink.
It implements "fast merging", i.e. copy-based merging that does not decompress and recompress pages.
The RNTupler merger is used by the `TFileMerger` and thus provides RNTuple merge support in `hadd` and `TBufferMerger`.

### RNTupleImporter
The RNTupleImporter creates RNTuple data sets from ROOT trees.

> It is part of the RNTupleUtil library.

Reviewer 2: ```suggestion
It is part of the `ROOTNTupleUtil` library.
```

### RNTupleInspector
The RNTupleInspector provides insights of an RNTuple, e.g. the distribution of data volume wrt. column types.

> It is part of the RNTupleUtil library.

Reviewer 2: ```suggestion
It is part of the `ROOTNTupleUtil` library.
```

On committing the cluster, all pages are sealed and sent to a _persistent sink_ in one go (vector write).
Pages are also reordered to ensure locality of pages of the same column.

### Reading Case
The reverse process is performed on reading (e.g. `RNTupleReader::LoadEntry()`, `RNTupleView` call operator).

By default, the page source uses an `RClusterPool` to asynchronously read-ahead data.
When a page of a certain cluster is required, the cluster pool reads pages of _active_ columns.
For instance, if only certain fields are used (e.g., through an imposed model), only the pages of columns connected to those fields are read.
Columns can be dynamically added (e.g. during event iteration, a new field view is created in a reader).
The page pool reads ahead a limited number of clusters given by the _cluster bunch size_ option (default = 1).
The read-ahead uses vector reads.
For the file backend, it additionally coalesces close read requests and uses uring reads when it is available.

The page source can be restricted to a certain entry range.
This allows for optimizing the page lists that are being read.

> And it allows for optimizing the cluster pool to not read-ahead beyond the limits.

Reviewer 2: ```suggestion
Additionally, it allows for optimizing the cluster pool to not read-ahead beyond the limits.
```

An S3 backend is planned.

Every new storage backend needs to define
  1) The RNTuple embedding: how are RNTuple data blobs stored, e.g. in keys of ROOT files, or in objects of object stores
  2) The RNTuple anchor: the initial link to the location of the header and footer (cf. format specification)
  3) A locator format: how are byte ranges addressed (e.g., through an offset in a file or an object ID)

That means that new backends are likely to have implications on the RNTuple format specification.

The page sources and sinks are ROOT internal classes.
They are not meant to be extended by users.

Multi-Threading
---------------

The following options exist in RNTuple for multi-threaded data processing.

### Implicit Multi-Threading
When `ROOT::EnableImplicitMT()` is used, RNTuple uses ROOT's task arena to compress and decompress pages.
That requires writes to be buffered and reads uses the cluster pool resp.
The RNTuple data source for RDataFrame lets RDataFrame full control of the thread pool.
That means that RDataFrame uses a separate data source for every thread, each of the data sources runs in serial mode.

### Concurrent Readers

> Multiple readers and read the same RNTuple concurrently as long as access to every individual reader is sequential.

Reviewer 3: This is not clear to me, maybe there are some words missing?

a brief serialization point handles the RNTuple meta-data updates and the reservation of disk space to write into.

Relationship to other ROOT components
-------------------------------------

The RNTuple classes have the following relationship to other parts of ROOT.

The RNTuple classes use core ROOT infrastructure classes, such as error handling and logging.
When necessary, RNTuple uses a `TFile` for reading and writing.
The cases of writing to a local file and reading from a local file, a file from XRootD or from HTTP, do _not_ require `TFile`.
For these cases, RNTuple depends on the `RRawFile` class and its XRootD and Davix plugins.

For user-defined classes as well as sets and maps, RNTuple uses `TClass`.
Simple types and other stdlib classes are natively supported and do not require dictionaries.

Integration to RDataFrame is provided through an RNTuple data source.
A universal RDataFrame constructor can create a data frame from either a TTree or an RNTuple with the same syntax.

The RBrowser uses RNTuple classes to display RNTuple dataset information.

Future Features
---------------

The following feature a planned for after the first RNTuple production version:

>   - RNTupleMT: advanced RNTupleReader that allows for free combination of chains and (indexed) friends

Reviewer 3: Is this the new "processor" that we talked about?

Author: Yep, should be called RNTupleProcessor
